### PR TITLE
ASM: Introduce labels and relocatable symbols

### PR DESCRIPTION
**Pull request by pr1metine**
See [this guide on ELF object writing](https://www.embecosm.com/appnotes/ean10/ean10-howto-llvmas-1.0.html#idp3613728).

# Todo
- Figure out how to conduct lit tests 🔥
- Figure out how to relocate symbols after emitting object file

### CODE REVIEW

**Review by indicatelovelace — COMMENTED** [review]
explain this weird jumpWithLabelRelocations.s
nop
nop
nop
...

**Comment by pr1metine** [comment]
I wanted to test symbol references for addresses higher than 0xFF. The `what` label points to 0x120. Therefore, `goto what` should be encoded to [0x20,0x29]. However, I encountered some problems when trying to emit correct machine code for this instruction. Thus, `jumpWithLabelRelocations.s` serves as a reference for future tests. In the future, we should somehow automate relocation testing.

File: llvm/test/MC/PICMid/jumpWithLabelRelocation.s [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    nop
+    nop
+    nop
+    nop
+    nop
+    nop
+what:
+    ADDLW 128
+    ADDWF 69, 1
+    goto what
```
Comment (pr1metine) [inline llvm/test/MC/PICMid/jumpWithLabelRelocation.s:294]: Basically this

**Review by indicatelovelace — APPROVED** [review]

**Comment by pr1metine** [comment]
Thx 🙏

**Pull request merged**